Description: We introduce a new preference that allows users to set a preferred
 plugin for a given mime-type.

 For example:

  pref ("modules.plugins.mimetype.application/x-shockwave-flash", "/usr/lib/firefox-3.0.1/plugins/libflashplayer.so")

 would make the flashplyer installed in that location the preferred one to use.

 In case the path is not valid, we just go ahead and search for the first match
 given the mime-type.
Author: Alexander Sack <asac@ubuntu.com>
Forwarded: no

Index: mozilla/dom/plugins/base/nsPluginHost.cpp
===================================================================
--- mozilla.orig/dom/plugins/base/nsPluginHost.cpp	2012-04-10 11:46:52.000000000 +0100
+++ mozilla/dom/plugins/base/nsPluginHost.cpp	2012-04-10 22:59:49.487175839 +0100
@@ -1558,7 +1558,35 @@
 
   LoadPlugins();
 
+  nsCAutoString mimetypePref("modules.plugins.mimetype.");
+  mimetypePref.Append(aMimeType);
+
+  nsAdoptingCString preferred = Preferences::GetCString(mimetypePref.get());
+
   nsPluginTag *plugin = mPlugins;
+  if(!preferred.IsEmpty()) {
+    nsCOMPtr<nsILocalFile> file;
+    nsresult rv = NS_NewNativeLocalFile(preferred, false, getter_AddRefs(file));
+    bool isAbsolute = NS_SUCCEEDED(rv);
+
+    while (plugin) {
+      if (((!isAbsolute && 0 == PL_strcasecmp(plugin->mFileName.get(), preferred.get())) ||
+           (isAbsolute && 0 == PL_strcasecmp(plugin->mFullPath.get(), preferred.get()))) &&
+          (!aCheckEnabled || plugin->IsEnabled())) {
+        PRInt32 mimeCount = plugin->mMimeTypes.Length();
+        for (PRInt32 i = 0; i < mimeCount; i++) {
+          if (0 == PL_strcasecmp(plugin->mMimeTypes[i].get(), aMimeType)) {
+            return plugin;
+          }
+        }
+      }
+      plugin = plugin->mNext;
+    }
+  }
+
+  // if there is no pref for this mime-type, or if the plugin named in pref
+  // isn't found, we pick the first that matches for this mime-type
+  plugin = mPlugins;
   while (plugin) {
     if (!aCheckEnabled || plugin->IsEnabled()) {
       PRInt32 mimeCount = plugin->mMimeTypes.Length();
